cirq/circuits/circuit.py

```diff
@@ -1767,12 +1766,10 @@ def on_stuck(bad_op):
 
 
 def _decompose_measurement_inversions(op: ops.Operation) -> ops.OP_TREE:
-    if ops.MeasurementGate.is_measurement(op):
-        gate = cast(ops.MeasurementGate, cast(ops.GateOperation, op).gate)
-        return [ops.X(q)
-                for q, b in zip(op.qubits, gate.invert_mask)
-                if b]
-
+    if ops.op_has_gate_type(op, ops.MeasurementGate):
```

Comment:
I would make this a method on `Operation` itself, e.g. `op.has_gate_is_instance(...)`.

I agree that it's a bit hacky, but given how often we do it it may be justified to work this way:

`isinstance(op.gate, ops.MeasurementGate)`

where `op.gate` is a property that returns None and `GateOperation` overrides this attribute to return its gate.

Reply:
I'm not following you.  Making a method in Operation that refers to something on GateOperation only seems....like a bad coupling?

Then you are advocating making gate a property on Operation?  Most of the time when you are asking if a gate is of a particular type you will need to cast because you want to use properties of that specific gate type (we avoid this when we can use a protocol, which is why they are nice).

Reply:
I guess it does couple.

Maybe instead of `op_has_gate_type` the method should be `op_gate_of_type(T) -> Optional[T]`.